### PR TITLE
Count-agnostic Pagination

### PR DESCRIPTION
**Pull request by afrasier**
Pre-merge checklist:

- [x] Backend review
- [x] Deploy to Sandbox backend ( @willkjackson let me know when it is available for use )
- [x] @ebdabbs has prepared front end PR to switch to the new pagination style
- [x] Front-end changes and back-end changes have been tested together on sandbox
- [ ] Final merge into dev

#### POST request preservation

When a user send a request, a new entry in the database is created. This entry (the model `ResponseCatalog` in `usaspending_api/common/models.py`) stores the request object (_minus_ any pagination parameters) and a checksum generated from the SHA-256 hash of the serialized request. 

This checksum is returned as the parameter `req`, for example: 

```
"page_metadata": { . . . },
"results": [ . . . ],
"req": "a29927e91743c6d40f7ca2817edf32af3b74bbd3e88d5028fbf8f0efad342c0d",
```

This parameter can then be used via GET or POST to re-run that request. If you generated the above checksum with a filter that only selected types of "A", hitting the same endpoint via GET (`/api/v1/awards/?req=a29927e91743c6d40f7ca2817edf32af3b74bbd3e88d5028fbf8f0efad342c0d`) will re-produce the same results by getting those filters from the database. 

Tests have been added to ensure this functionality is consistent, and works on all endpoints. Additionally, they track the last_accessed time so we can (if the table gets to large) cull rarely-used requests. 

This feature has an added bonus of being a major step forward on search-page sharing for the front-end. 

#### Count-agnostic pagination
Our previous pagination was done using a `PagePaginator` which required us to count large sets of data (a problem soon to be exacerbated as we load more data). To create a count-agnostic paginator using the same field names (`page` and `limit`), I created a limit/offset paginator that incorporated the traversal methods of a cursor paginator. 

Thus, a user still requests a page using `page` and `limit`, but will no longer be provided with the total count. Instead, they only review `page_metadata`. For example:

```
"page_metadata": {
        "page": 2,
        "has_next_page": true,
        "next": "http://localhost:8000/api/v1/awards/?limit=100&page=3&req=e359928279a3a43990cd2819fd69d355d9b7e92256700b62db46ecba0c88b556",
        "previous": "http://localhost:8000/api/v1/awards/?limit=100&page=1&req=e359928279a3a43990cd2819fd69d355d9b7e92256700b62db46ecba0c88b556"
    },
```

Here we see the page metadata for a request to `/api/v1/awards/`. It tells us we are on page 2, that a next page exists, and provides us with easy to use links for the previous and next pages. Note that these pages use the `req` argument mentioned in the previous section, so we can now navigate large datasets using GET once the first set of filters has been POSTed.

This also allows us to use templated pagination navigation on our API auto-doc pages:
screen shot 2017-03-29 at 5 06 20 pm
(Note the "previous" and "next" buttons)

### What's changed?

* Updated all Data and Aggregate endpoints to use the new Paginator following the DRF Framework more strictly (Autocomplete endpoints don't paginate)
* Removed unused code: `UniqueValueHandler`, `DataQueryHandler`, `ResponsePaginator`, `ResponseMetadataMixin`, `AggregateView` 
* Former `AggregateView` endpoints now use `DetailViewSet` with a serializer of `AggreagteSerializer` instead
* Added tests for new pagination and request slugs
* Updated tests where applicable 
* Updated travis configuration to use Postgres 9.4 (due to the use of JSONField)
* Updated documentation for new pagination fields and request slugs

### CODE REVIEW

File: usaspending_api/common/api_request_utils.py [hunk 1]
```diff
@@ -629,19 +629,31 @@ def get_paged_data(data_set, page=1, page_limit=100, request_parameters={}):
...
+
+        # Let's try an offset-limit nonsense
+        limit = page_limit * page
+        offset = page_limit * (page - 1)
+
+        # Does next page exist?
+        next_limit = page_limit * (page + 1)
+        next_offset = page_limit * page
+
+        print(data_set[next_offset:next_limit].count())
```
Comment (bsweger) [inline usaspending_api/common/api_request_utils.py]: Is this a leftover debug line?

File: usaspending_api/common/mixins.py [hunk 2]
```diff
@@ -215,62 +215,6 @@ def build_response(self, request, *args, **kwargs):
         return AutoCompleteHandler.handle(queryset, params, serializer)
 
 
-class ResponseMetadatasetMixin(object):
```
Comment (bsweger) [inline usaspending_api/common/mixins.py]: Happy to say goodbye to this one!

File: usaspending_api/common/pagination.py [hunk 3]
```diff
@@ -1,40 +1,117 @@
...
+        # If our next page has a count > 0, return true
+        return (bool(queryset[(self.offset+self.limit):(self.offset+self.limit*2)]))
+
+    def get_limit(self, request):
+        # Check both POST and GET parameters for limit
+        request_parameters = {**request.data, **request.query_params}
+
+        if 'limit' in request_parameters:
+            # We need to check if this is a list due to the fact we support both POST and GET
+            l = request_parameters['limit']
```
Comment (bsweger) [inline usaspending_api/common/pagination.py]: I'm curious--why would `limit` be an integer in one type of request and a list in another type of request?

File: usaspending_api/common/pagination.py [hunk 4]
```diff
@@ -1,40 +1,117 @@
...
+                l = l[0]
+            return min(int(l), self.max_page_size)
+        else:
+            return self.page_size
+
+    def get_page(self, request):
+        # Check both POST and GET parameters for page
+        request_parameters = {**request.data, **request.query_params}
+        if 'page' in request_parameters:
+            # We need to check if this is a list due to the fact we support both POST and GET
```
Comment (bsweger) [inline usaspending_api/common/pagination.py]: Same question as for `limit` above--does it make sense to just accept integers for these parameters, regardless of request type?

File: usaspending_api/common/tests/test_endpoints.py [hunk 5]
```diff
@@ -71,6 +71,9 @@ def test_endpoints(endpoint_data, client):
         # be a multi-tiered nest of lists and objects, and can also be OrderedDicts
         # and ResultLists, which don't play nice with the native equality checks
         # TESTING TODO: I bet I can beat this
+        print(url)
```
Comment (bsweger) [inline usaspending_api/common/tests/test_endpoints.py]: Just tagging some stray debug statements...

File: usaspending_api/common/views.py [hunk 6]
```diff
@@ -6,76 +6,14 @@
...
 from usaspending_api.common.models import RequestCatalog
-from usaspending_api.common.mixins import AggregateQuerysetMixin, AutocompleteResponseMixin
+from usaspending_api.common.mixins import AutocompleteResponseMixin, FilterQuerysetMixin
 
 from usaspending_api.common.exceptions import InvalidParameterException
 
 import logging
 
 
-class AggregateView(AggregateQuerysetMixin,
```
Comment (bsweger) [inline usaspending_api/common/views.py]: Another piece of code that I'm happy to bid farewell--nicely done!

File: usaspending_api/common/views.py [hunk 7]
```diff
@@ -105,29 +43,33 @@ class DetailViewSet(viewsets.ReadOnlyModelViewSet):
...
     def list(self, request, *args, **kwargs):
         try:
+            # First, get the req object from RequestCatalog
             created, self.req = RequestCatalog.get_or_create_from_request(request)
-            response = self.build_response(
-                request, queryset=self.get_queryset(), serializer=self.get_serializer_class())
-            status_code = status.HTTP_200_OK
+            # Pass this to the paginator
+            self.paginator.req = self.req
+            # Get the queryset (the view will handle filtering and ordering)
```
Comment (bsweger) [inline usaspending_api/common/views.py]: I'm a bit confused here re: "the view will handle filtering and ordering". Isn't this the view?

File: usaspending_api/financial_activities/views.py [hunk 8]
```diff
@@ -1,13 +1,12 @@
...
 from usaspending_api.financial_activities.serializers import FinancialAccountsByProgramActivityObjectClassSerializer
-from usaspending_api.common.mixins import FilterQuerysetMixin, ResponseMetadatasetMixin, SuperLoggingMixin
+from usaspending_api.common.mixins import FilterQuerysetMixin, SuperLoggingMixin
 from usaspending_api.common.views import DetailViewSet
 
 
 class FinancialAccountsByProgramActivityObjectClassListViewSet(
         SuperLoggingMixin,
         FilterQuerysetMixin,
-        ResponseMetadatasetMixin,
```
Comment (bsweger) [inline usaspending_api/financial_activities/views.py]: 😀

File: usaspending_api/references/views.py [hunk 9]
```diff
@@ -2,27 +2,37 @@
...
+from usaspending_api.common.api_request_utils import GeoCompleteHandler, AutoCompleteHandler
 from usaspending_api.references.models import Location, Agency, LegalEntity, CFDAProgram
 from usaspending_api.references.serializers import LocationSerializer, AgencySerializer, LegalEntitySerializer, CfdaSerializer
 from usaspending_api.common.mixins import FilterQuerysetMixin, SuperLoggingMixin
 from usaspending_api.common.views import DetailViewSet, AutocompleteView
 import json
 
 
-class LocationEndpoint(APIView):
+class LocationEndpoint(SuperLoggingMixin,
```
Comment (bsweger) [inline usaspending_api/references/views.py]: Nice! Glad to see more views switching from `APIView` to a Viewset!

File: usaspending_api/references/views.py [hunk 10]
```diff
@@ -2,27 +2,37 @@
...
+    def get_queryset(self):
+        """Return the view's queryset."""
+        queryset = Location.objects.all()
+        queryset = self.serializer_class.setup_eager_loading(queryset)
+        filtered_queryset = self.filter_records(self.request, queryset=queryset)
+        ordered_queryset = self.order_records(self.request, queryset=filtered_queryset)
+        return ordered_queryset
+
+
+class LocationGeoCompleteEndpoint(APIView):
```
Comment (bsweger) [inline usaspending_api/references/views.py]: So much nicer as separate classes--thanks!

File: usaspending_api/api_docs/markdown/using_the_api.md [hunk 11]
```diff
@@ -251,6 +247,24 @@ Field names support Django's foreign key traversal; for more details on this see
...
+```
+{
+  "req": "abcd"
+}
+```
+
+Do note that this _ignores_ pagination variables, so you can request a different page from the same request without changing your `req`. For example:
+
+`/api/v1/awards/?req=abcd&page=2`
+
```
Comment (bsweger) [inline usaspending_api/api_docs/markdown/using_the_api.md]: What do you think about supplementing this with an example of passing a page number to a `POST` request?

**Comment by bsweger** [comment]
@afrasier One other thought, to follow-up on your idea from yesterday....would you mind setting up a quick synchronous demo of the new pagination features? This is much more than a new way to paginate--it's setting us up for links directly to search results, etc and maybe even move to a place where we can operate mostly via `GET`. Seems worthwhile to discuss these larger implications with the team.

File: usaspending_api/common/models.py [hunk 12]
```diff
@@ -19,14 +25,74 @@ class Meta:
...
+        # Delete any pagination data
+        for item in ["page", "limit"]:
+            if data.get(item, None):
+                del data[item]
+            if query_params.get(item, None):
+                del query_params[item]
+
+        json_request = {"data": data, "query_params": query_params}
+
+        print(json_request)
```
Comment (afrasier) [inline usaspending_api/common/models.py]: Yes, will remove

File: usaspending_api/common/models.py [hunk 13]
```diff
@@ -19,14 +25,74 @@ class Meta:
...
+    request = JSONField(null=False, help_text="The serialized form of the POST request")
     checksum = models.CharField(max_length=256, unique=True, null=False, db_index=True, help_text="The SHA-256 checksum of the serialized POST request")
     create_date = models.DateTimeField(auto_now_add=True, blank=True, null=True)
     last_accessed = models.DateTimeField(auto_now_add=True, blank=True, null=True)
 
-    def save(self, *args, **kwargs):
-        self.checksum = hashlib.sha256(self.request).hexdigest()
-        super(RequestSlugs, self).save(*args, **kwargs)
+    @staticmethod
+    def get_or_create_from_request(request):
```
Comment (afrasier) [inline usaspending_api/common/models.py]: Will do, good point

File: usaspending_api/common/models.py [hunk 14]
```diff
@@ -19,14 +25,74 @@ class Meta:
...
+
+        print(json_request)
+
+        # Make the checksum
+        checksum = hashlib.sha256(json.dumps(json_request).encode('utf-8')).hexdigest()
+
+        # See if the checksum exists, and return that catalog. Otherwise, create it
+        created = False
+        request_catalog = None
+        try:
```
Comment (afrasier) [inline usaspending_api/common/models.py]: Possibly, I'll try it out

File: usaspending_api/common/models.py [hunk 15]
```diff
@@ -19,14 +25,74 @@ class Meta:
...
+            request_catalog.save()
+            created = False
+        except ObjectDoesNotExist:
+            request_catalog = RequestCatalog.objects.create(request=json_request, checksum=checksum)
+            created = True
+
+        return created, request_catalog
+
+    def merge_requests(self, request):
+        '''Merges a request with the stored values (i.e. taking pagination data from incoming request)'''
```
Comment (afrasier) [inline usaspending_api/common/models.py]: This is actually not even used, ignore everything about it. It's just leftovers I forgot to remove.

File: usaspending_api/common/views.py [hunk 16]
```diff
@@ -81,7 +82,7 @@ class AutocompleteView(AutocompleteResponseMixin,
     def post(self, request, *args, **kwargs):
         try:
             response = self.build_response(
-                self.request, queryset=self.get_queryset(), serializer=self.serializer_class)
+                request, queryset=self.get_queryset(), serializer=self.serializer_class)
             status_code = status.HTTP_200_OK
         except InvalidParameterException as e:
             response = {"message": str(e)}
```
Comment (afrasier) [inline usaspending_api/common/views.py]: Yeah, if you send a checksum that doesn't exist it generates an InvalidParameterException

File: usaspending_api/common/views.py [hunk 17]
```diff
@@ -111,8 +112,10 @@ class DetailViewSet(viewsets.ReadOnlyModelViewSet):
     @cache_response()
     def list(self, request, *args, **kwargs):
         try:
+            created, req = RequestCatalog.get_or_create_from_request(request)
             response = self.build_response(
-                self.request, queryset=self.get_queryset(), serializer=self.get_serializer_class())
+                request, queryset=self.get_queryset(), serializer=self.get_serializer_class(),
```
Comment (afrasier) [inline usaspending_api/common/views.py]: Yes, it will work even if the original request is a GET.

File: .travis.yml [hunk 18]
```diff
@@ -5,6 +5,9 @@ python:
 
 services: postgresql
 
+addons:
+  postgresql: "9.4"
```
Comment (afrasier) [inline .travis.yml]: Yep! It's required for JSONField. Travis just defaults to 9.1

File: usaspending_api/api_docs/markdown/using_the_api.md (hunk 11 above)
Comment (afrasier) [inline usaspending_api/api_docs/markdown/using_the_api.md]: Good idea! I'm also going to add a specific tutorial about using the request checksums, etc.

File: usaspending_api/common/pagination.py (hunk 3 above)
Comment (afrasier) [inline usaspending_api/common/pagination.py]: I'm not entirely sure. I'm pretty sure django supports multiples of the same key via GET (which is when this is a list). So you could do `?page=1&page=2`, hence why the query_params are lists.

File: usaspending_api/common/pagination.py (hunk 4 above)
Comment (afrasier) [inline usaspending_api/common/pagination.py]: We are only accepting integers, it's just that GET requests are sending lists. See above.

File: usaspending_api/common/tests/test_endpoints.py (hunk 5 above)
Comment (afrasier) [inline usaspending_api/common/tests/test_endpoints.py]: Cheers! :tada:

File: usaspending_api/common/tests/test_request_catalog.py [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+    # Check that a post query for this request checksum returns the same thing
+    resp = client.post(
+        '/api/v1/awards/',
+        content_type='application/json',
+        data=json.dumps({"req": checksum}))
+
+    assert resp.status_code == status.HTTP_200_OK
+    assert results == resp.data['results']
+
+    # Check that a post query returns the same thing
```
Comment (afrasier) [inline usaspending_api/common/tests/test_request_catalog.py]: That is correct, will fix!

File: usaspending_api/common/tests/test_request_catalog.py [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+    resp = client.get('/api/v1/awards/?req=' + checksum)
+
+    assert resp.status_code == status.HTTP_200_OK
+    assert results == resp.data['results']
+
+    # check that a no-good checksum returns 500
```
Comment (afrasier) [inline usaspending_api/common/tests/test_request_catalog.py]: Copy and pasting is bad for the heart, it seems.

**Comment by afrasier** [comment]
Thanks for the review - I'll be addressing the comments as noted tomorrow, and should have fixes in then.

**Comment by afrasier** [comment]
No problem with a demo - just set something up on the calendar and I'd be glad to give a tour.

File: usaspending_api/common/views.py (hunk 7 above)
Comment (afrasier) [inline usaspending_api/common/views.py]: I'll update to clarify, it means the view for the app, not the viewset here (i.e. get_queryset handles it)

File: usaspending_api/common/tests/test_request_catalog.py [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+import json
+from datetime import date
+
+from model_mommy import mommy
+from rest_framework import status
+
+from usaspending_api.awards.models import Award
+
+
+@pytest.fixture
```
Comment (afrasier) [inline usaspending_api/common/tests/test_request_catalog.py]: It exists! Scroll down! :)

File: usaspending_api/common/pagination.py [hunk 22]
```diff
@@ -50,7 +51,8 @@ def get_limit(self, request):
             l = request_parameters['limit']
             if isinstance(l, list):
                 l = l[0]
-            return min(int(l), self.max_page_size)
+            # This will ensure our limit is bounded by [1, max_page_size]
```
Comment (bsweger) [inline usaspending_api/common/pagination.py]: Thanks for the comment!

File: usaspending_api/common/models.py [hunk 23]
```diff
@@ -33,6 +33,14 @@ class RequestCatalog(models.Model):
 
     @staticmethod
     def get_or_create_from_request(request):
+        """
```
Comment (bsweger) [inline usaspending_api/common/models.py]: Thanks for adding this docstring!

**Comment by bsweger** [comment]
@afrasier Approved--thanks for the updates. Will let you handle coordination with front-end and do the final merge to dev.

**Pull request merged**